### DISCUSSION TITLE
Feature Request: add build time transformations to useStaticQuery

### DISCUSSION THREAD

**Discussion by khinshankhan** [gatsbyjs/gatsby#34198] [Ideas / Feature Requests]
This is an enhancement to the `useStaticQuery` hook. According to this <https://www.gatsbyjs.com/docs/static-vs-normal-queries/> the queries results are saved into json files. I think it'd be neat and doable to pass in transformations to the hook, eg

```js
  useStaticQuery(
    graphql`
      {
        allMdx {
          edges {
            node {
              id
            }
          }
        }
      }
    `,
    (result) => result.allMdx.edges.map((node) => ({ ...node.node }))
  )
```

This is a rather simplified example, there are definitely much more complex uses of this, but the benefit is that now these transformations occur at build time rather than during each user request. 

Data transformations could make data easier to deal with but the real benefit I see it they can also change the data for different types of access that a raw query may not be able to readily support. Eg if an array there's an array of tags/ categories, developers may want constant lookup for them. A good way would be to use a set or map structure, but graphql doesn't support readily those data types, especially since every field must be specified. An array would be the best type for the query, but a data transformation could make it so `["a", "b"]` => `{a: true, b: true}`, providing the constant lookup.

**Comment by violy** [comment]
Does `useStaticQuery` (and `<StaticQuery>`) should override datas with transformations… I don't think so. 
Solving complexity and performance is difficult, and needs trade-offs. But adding layers of functions is probably not the best way to solve it, IMO. 

GraphQL is _by design_ made to get optimized and _view-based_ datas. 
The problem exposed here, is about **GraphQL limitations** — in practice, more than in theory.

**Reply by khinshankhan** [reply]
I think it's misunderstood what the request is. I think it's fine to query the data as is with graphql limitations, but when _caching_ the result, the data could be transformed. 

Perhaps I misunderstood the hashed file part, but this [link](https://www.gatsbyjs.com/docs/static-vs-normal-queries/#adding-imports-for-page-data) indicated to me that that the results are saved in json files. Perhaps the confusing is the useStaticQuery may be hashing _only_ the query (?) so maybe a different hook needs to be introduced. The new hook may call `useStaticQuery` under the hood to generate the hashed file for the query, and then the wrapping hook creates a hashed file, separate from the initial query but has the transformation.

**Reply by violy** [reply]
As mentioned, `useStaticQuery` results datas is stored (not _cached_) in plain `json` files. ok. 

You want to *transform* (and optimize) datas, just *after* the graphql return, before stored… ok. 

I just think that you (in fact, via graphql mutations) should make this transforms (during query run) but it's might be a painful way. here's the difficulty : how to add custom mutation to get most optimized datas, via `graphql` ? am i right ?

**Reply by khinshankhan** [reply]
Yes. But I don't particularly care about optimization, I just think allowing us to shape the data is powerful.

Also, I'm still kind of new to graphql but aren't graphql mutations write operations? Does the filesystem + redux create a "database" and graphql mutations would modify the stored files? Is there documentation for that?
